gtsam/basis/Chebyshev.h

```diff
@@ -0,0 +1,92 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Chebyshev.h
+ * @brief Chebyshev basis decompositions
+ * @author Varun Agrawal, Frank Dellaert
+ * @date July 4, 2020
+ */
+
+#pragma once
+
+#include <gtsam/base/Manifold.h>
+#include <gtsam/basis/Basis.h>
+
+#include <unsupported/Eigen/KroneckerProduct>
+
+namespace gtsam {
+
+/**
+ * Basis of Chebyshev polynomials of the second kind.
+ * https://en.wikipedia.org/wiki/Chebyshev_polynomials#Second_kind
+ * These are typically denoted with the symbol U_n, where n is the degree.
+ * The parameter N is the number of coefficients, i.e., N = n+1.
+ * In contrast to the templates in Chebyshev2, the classes below specify
+ * basis functions, weighted combinations of which are used to approximate
+ * functions. In this sense, they are like the sines and cosines of the Fourier
+ * basis.
+ */
+struct Chebyshev2Basis : Basis<Chebyshev2Basis> {
+  using Parameters = Eigen::Matrix<double, -1, 1 /*Nx1*/>;
+
+  /**
+   *  Evaluate Chebyshev Weights on [-1,1] at any x up to order N-1 (N values)
+   */
+  static Weights CalculateWeights(size_t N, double x);
+};
+// Chebyshev2Basis
+
+/**
+ * Basis of Chebyshev polynomials of the first kind
+ * https://en.wikipedia.org/wiki/Chebyshev_polynomials#First_kind
+ * These are typically denoted with the symbol T_n, where n is the degree.
+ * The parameter N is the number of coefficients, i.e., N = n+1.
+ */
+struct Chebyshev1Basis : Basis<Chebyshev1Basis> {
+  using Parameters = Eigen::Matrix<double, -1, 1 /*Nx1*/>;
+
+  /**
+   *  Evaluate Chebyshev Weights on [-1,1] at x up to order N-1 (N values)
+   */
+  static Weights CalculateWeights(size_t N, double x);
```

Comment:
Do CRTP requirements need to be set? It's not the same as inheritance so adding them even when they don't make sense in the context isn't the right thing to do.

Reply:
Oh I guess I just assumed it would be best practice but maybe not?  I'm not that familiar.

I guess I was just thinking that, as a user, I was looking at `Basis` for all the functions I could use and wanted a Chebyshev1 EvaluatorFunctor with interval from 0 to 1 but then I got a compile error and was surprised (because that constructor existed in Basis but not for Chebyshev1).

Maybe a thought: that if it doesn't satisfy a CRTP requirement then maybe the CRTP base should not include that? (e.g. split into 2 bases to separate out functionality that isn't common)?  I think actually in this case there is a reasonable definition of the domain [a, b] for all inherited types (including Fourier, see comment below).  For derivatives, one option could be placing derivatives in a separate CRTP base?

Again, I don't know as much about this but just offering possible suggestions.

Reply:
Just thought of another possible suggestion: variadic function templates?  This could maybe mitigate the confusion of a definition that exists in Basis but that cannot be used for certain basis functions, since the variadic template makes semantically clear that it depends on the derived implementation what type of calls are possible.

Reply:
I see what you mean. Giving users the ability to restrict the domain is easy so I'll add it.

Reply:
I don't see any math for how to restrict the limits of the polynomials for Cheb1 and Cheb2, but I'm putting in the limits for consistency.

Reply:
This actually goes back to the limit not making sense since for Chebyshev1 polynomials, a factor with `graph.emplace_shared<EvaluationFactor<Chebyshev1Basis>>(key, x(i), model, N, t(i), -1.0, 1.0);` essentially does nothing compared to `graph.emplace_shared<EvaluationFactor<Chebyshev1Basis>>(key, x(i), model, N, t(i));`.

This is esoteric knowledge however.